pkg/controller/direct/bigquerydatatransfer/bigquerydatatransferconfig_controller.go

```diff
@@ -0,0 +1,351 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bigquerydatatransfer
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/apis/bigquerydatatransfer/v1alpha1"
+	refv1beta1 "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
+
+	gcp "cloud.google.com/go/bigquery/datatransfer/apiv1"
+	bigquerydatatransferpb "cloud.google.com/go/bigquery/datatransfer/apiv1/datatransferpb"
+	"google.golang.org/api/option"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	ctrlName      = "bigquerydatatransfer-controller"
+	serviceDomain = "//bigquerydatatransfer.googleapis.com"
+)
+
+func init() {
+	registry.RegisterModel(krm.GroupVersionKind, NewModel)
+}
+
+func NewModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &model{config: *config}, nil
+}
+
+var _ directbase.Model = &model{}
+
+type model struct {
+	config config.ControllerConfig
+}
+
+func (m *model) client(ctx context.Context) (*gcp.Client, error) {
+	var opts []option.ClientOption
+	if m.config.UserAgent != "" {
+		opts = append(opts, option.WithUserAgent(m.config.UserAgent))
+	}
+	if m.config.HTTPClient != nil {
+		opts = append(opts, option.WithHTTPClient(m.config.HTTPClient))
+	}
+	if m.config.UserProjectOverride && m.config.BillingProject != "" {
+		opts = append(opts, option.WithQuotaProject(m.config.BillingProject))
+	}
+
+	gcpClient, err := gcp.NewRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building bigquerydatatransfer client: %w", err)
+	}
+	return gcpClient, err
+}
+
+func (m *model) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.BigQueryDataTransferConfig{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	// Get ResourceID
+	resourceID := direct.ValueOf(obj.Spec.ResourceID)
```

Comment:
Did you miss a case that metadata.name --> resourceID if `spec.resourceID` is not given? Or is that not a valid case here? I mean the case that KCC *acquire* a GCP resource that matches the `metadata.name`.

Reply:
For resource with service generated ID, I am not sure how useful is `metadata.name`. It will be ignored during GCP resource creation, and later it cannot be used to match (acquire) the GCP resource. To acquire the GCP resource, we have to give GCP API the service generated ID.

Reply:
From my tests, the generated ID is something like "66e60359-0000-2d33-8bc7-c82add80addc". I think k8s name needs to start and end with an alphanumeric character?

Besides, when user try to acquire a GCP resource, I prefer them to explicitly set `spec.ResourceID`.

Reply:
kk, looks like all BigQuery service-generated ID is all using the UUID format. Then I'm good with this.